### ISSUE TITLE
pull_exn throws an exception on 32-bit system (rPi Debian Wheezy)

### ISSUE THREAD

**Issue by gregtatcam** [mirage/ocaml-git#97]
The trace:

Invalid_argument("String.create") Raised at file "map.ml", line 117, characters 16-25
Called from file "src/core/lwt.ml", line 161, characters 4-40
Re-raised at file "src/core/lwt_pqueue.ml", line 79, characters 23-32
Called from file "src/core/lwt_pqueue.ml", line 87, characters 14-26
Re-raised at file "src/core/lwt_pqueue.ml", line 79, characters 23-32
Called from file "src/core/lwt_pqueue.ml", line 87, characters 14-26

Using Git 1.5.2 and Irmin 0.9.7. The issue seems to be related to the repo's size. Repo under 25MB works fine. String.create has a limit of about 17MB on 32-bit system.
Maybe there is a way to improve git implementation since the conventional git command line works fine?

**Comment by samoht** [comment]
There is indeed a 16M limit for string sizes in OCaml 32b. I'll try to think of a workaround but that could take some time.

**Comment by gregtatcam** [comment]
Thanks!
Let me know if I can help. I need to do 2nd year report by the end of July but then I’ll hopefully have a bit of a break and might be able to spare a couple of weeks.

**Comment by samoht** [comment]
Actually I think that's not very difficult to fix. `IO.read_all` should return a list of chunks instead of a big string and that should solve the problem. I'm bit reluctant to do the change today as it might break at other places and I don't have time to test on 32b system but I can have a look at it before the end of the month.

**Comment by gregtatcam** [comment]
Great! I’ll be happy to test on rPi.

**Comment by samoht** [comment]
I've pushed a tentative fix in master. I'm installing a 32b compiler to test it.

**Comment by gregtatcam** [comment]
Thanks! I’ll test it on rPi.

**Comment by gregtatcam** [comment]
It still crashes, the same error.

**Comment by samoht** [comment]
I've just merged the patch into master, did you try with the right branch?

**Comment by gregtatcam** [comment]
I thought you already had it in the master. Let me try again…

**Comment by samoht** [comment]
ha yes, sorry I wasn't very clear. Let me know how it works, I wasn't able to compile a 32bit compiler on my machine... I'll make a 1.6.0 very soon anyway.

**Comment by gregtatcam** [comment]
Still the same crash. 
I clone ocaml-git, configure with —enable-unix and —disable-mirage; build; reinstall; then reinstall irmin; then rebuild my app. Does this sound right or am I missing anything?

**Comment by samoht** [comment]
How did you reinstall irmin? Easier way to test `ocaml-git`'s master is to do `opam pin add git --dev` and then you can install irmin as usual (you might need the dev version of Irmin as well...). Anyway, I'll cut new releases that will be easier to test for you.

**Comment by samoht** [comment]
You can also just test with `ogit` by doing `ogit clone <the repo>`

**Comment by gregtatcam** [comment]
I just did ‘opam reinstall irmin’

**Comment by samoht** [comment]
If you didn't pin `git` to your local Git repository, it will just pick the released version of `ocaml-git`. But if you have compiled the dev version of `ocaml-git`, just run `<path>/ogit.native clone <big-repo>`

**Comment by gregtatcam** [comment]
I did the pin and tried to ‘ogit clone’ - same error.

**Comment by samoht** [comment]
Fixed in [1.6.2](https://github.com/mirage/ocaml-git/releases/tag/1.6.2)
